book/src/recursive/negative_cycles.md

```diff
@@ -0,0 +1,256 @@
+
+If, after we try to prove `G1` and get back error, we later try to prove `G2`,
+we need to prove `not { G1 }`. Thus we recursively try to prove `G1`, which
+encounters a cached result (error). Because we hit a cached result, there is no
+negative cycle encountered. This `G1` is not provable, we conclude that `not {G1
+}` is true, and hence `G2` is true.
+
+You can see that the "cache" here plays the role of the "sets of true/false
+facts" that we talked about in the mathematical WFS definition. We also use the
+cache to define the "strata" between goals. Once something is cached, it means
+that we were able to compute a final result that was dependent purely on things
+already cached (and hence on a lower strata) or which is independent of any
+negative cycle (because negative cycles always give ambiguous results). Anything
+that we can then prove based on that cached reuslt must be in some other strata.
+
+## An example of a negative cycle that is ambiguous
+
+It's interesting to note that the previous example was only provable
+because the goal `G3` had no applicable clauses. Consider this variation:
+
+```notrust
+G1 :- not { G2 }, G3.
+G2 :- not { G1 }.
+G3.
+```
+
+Here, the goal `G3` can be proven -- this means that both `G1` and `G2` are
+ambiguous and can never be proven true or false. How would that play out
+in our solver?
+
+* To start, proving G1 would find that `not { G2 }` is ambiguous (as before),
+  and `G3` is true. That means the overall result is ambiguous, which would be
+  cached. As before, the result for `G2` would not be cached, as it is a
+  participant in a negative cycle.
+* If we go to recompute `G2`, this time we would find that `G1` is ambiguous,
+  and hence `not { G1 }` is also ambiguous.
+
+It's somewhat interesting to see why ambiguity arises here. The problem is
+that there are multiple consistent sets of answers in this scenario, and there
+is no good way to choose between them.
+
+* We could say that `G2` is *false*, in which case `G1` would be true.
+* We could say that `G1` is *false*, in which case `G2` would be true.
+
+These results would be "consistent", but we have no basis on which to say
+that `G1` or `G2` is false, and no real way to choose which of those two
+answers is "correct". Hence the result is considered ambiguous. 
+
+(There are alternative formulations, such as [Answer Set Programming][ASP], that
+can accommodate situations like this and give back the full set of
+possibilities.)
+
+[ASP]: https://en.wikipedia.org/wiki/Answer_set_programming
+
+## An example of a negative cycle that is provable
+
+The only way to "positively prove" something involving a negative cycle is for
+it to have multiple clauses, such that one clause may be ambiguous due to a
+cycle, but other clauses are fully provable, as in this example:
+
+```notrust
+G1 :- not { G2 }, G3.
+G1.
+G2 :- not { G1 }.
+```
+
+Here we have a negative cycle for `G1`, but only through one of the two
```

Comment:
> The other clause will give back a positive result. 

I might be missing something by how would that be. Isn't this other clause G3. If so how does it give a positive result if there is no rule for it?

Reply:
Sorry, that's confusing. G3 is the "other subgoal". The other *clause* is the clause `G1.`, which is trivially true.

Reply:
I'll try to rework the test to make it clearer at some point.

Reply:
I think the 'other clause' here is `G1.`

Reply:
Ah okey, I got it